src/navier.f90

```diff
@@ -1280,5 +1289,266 @@ subroutine tbl_flrt (ux1,uy1,uz1)
+        do k=1,xsize(3)
+            zm=dz*real(xstart(3)-1+k-1,mytype)-zc
+            r=sqrt(ym*ym+zm*zm)
+            do i=1,xsize(1)
+                if (r.le.ra.and.ep(i,j,k).eq.0) then
+                    ux(i,j,k)=ux(i,j,k)+(ub_constant-qmm)
+                endif
+            enddo
+        enddo
+    enddo
+
+    !Check new bulk velocity
+    qmm     = zero
+    ncountt = zero
+    qm      = zero
+    ncount  = zero
+    do k=1,xsize(3)
+        zm=dz*real(xstart(3)-1+k-1,mytype)-zc
+        do j=1,xsize(2)
+            if (istret.eq.0) ym=real(j+xstart(2)-1-1,mytype)*dy-yc
+            if (istret.ne.0) ym=yp(j+xstart(2)-1)-yc
+            r=sqrt(ym*ym+zm*zm)
+            do i=1,xsize(1)
+                if (r.le.ra.and.ep(i,j,k).eq.0) then
+                    qm=qm+ux(i,j,k)
+                    ncount=ncount+one
+                else
+                    !Cancel solid zone (ra <= r <= ra+wt)
+                    !and buffer zone (r > ra+wt)
+                    ux(i,j,k)=zero
+                    uy(i,j,k)=zero
+                    uz(i,j,k)=zero
+                endif
+            enddo
+        enddo
+    enddo
+    call MPI_ALLREDUCE(qm,qmm,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+    call MPI_ALLREDUCE(ncount,ncountt,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+    qmm=qmm/ncountt
+    if ((nrank==0).and.(mod(itime, ilist)==0)) print *,'    Bulk velocity  after',qmm
+    !
+    return
+    !
+  end subroutine pipe_bulk_u
+  !********************************************************************
+  !
+  subroutine pipe_bulk_phi(phi,ux,ep,is,phib_constant)
+  !
+  !********************************************************************
+
+    use decomp_2d
+    use decomp_2d_poisson
+    use variables
+    use param
+    use ibm_param, only: ra
+    use MPI
+
+    implicit none
+    !INPUTS
+    real(mytype),intent(inout),dimension(xsize(1),xsize(2),xsize(3))  :: phi,ux
+    real(mytype),intent(in   ),dimension(xsize(1),xsize(2),xsize(3))  :: ep
+    real(mytype),intent(in   )                                        :: phib_constant !bulk temperature value
+    !LOCALS
+    real(mytype)                                                      :: qv,qvm,qm,qmm !volumetric averaged values
+    real(mytype)                                                      :: ym,zm,yc,zc,r
+    real(mytype)                                                      :: ncount,ncountt
+    real(mytype)                                                      :: phi_out
+    integer                                                           :: is,j,i,k,code
+    character(len=30)                                                 :: filename
+    integer, allocatable, save, dimension(:)                          :: local_io_unit
+
+    if (iscalar.eq.0) return
+
+255 format('Tb_',I2.2,'.dat')
+256 format(' Scalar:                       #',I2)
+
+    ! Safety check
+    if (is<1.or.is>numscalar) return
+
+    if (itime.eq.ifirst.and.nrank.eq.0) then
```

Comment:
Same issue here but it is harder to fix. It is important to make sure the open statement is issued only once by rank 0

Reply:
A solution could be to set local_io_unit to -A when it is allocated. Then test `local_io_unit(is)` against -1. If it is -1, then the open statement should occur

Reply:
In ffc0c3c6b1f583d70523b2fea8ce541392590d47, I did as you've suggested, the problem is solved. Thanks again.